cilium-cli: Support testing with multiple parallel curl requests

jrajahalme opened the pull request:
Add new `--curl-parallel` option to `cilium connectivity test` to allow testing with multiple parallel curl requests to add stress to agent/proxy synchronization. This is opt-in, so by default only one request will be sent by curl commands. Takes the number of concurrent requests, e.g., `cilium connectivity test --curl-parallel=3`

Add new agent command line option `--proxy-max-concurrent-retries` (default 128, helm value `envoy.maxConcurrentRetries`). This is an increase to the Envoy default of only 3 on each upstream cluster. Higher number is warranted due to the default maximum number of upstream connections being 1024, and the fact that Cilium configures upstream clusters that are shared for all local endpoints. The new default is applied also to Envoy Clusters configured via CEC/CCEC CRDs.

Finally, expose the agent command line option `--http-retry-count` (default 3) via a new helm value `envoy.httpRetryCount` to make tuning this easier.

```release-note
Cilium-cli connectivity test now supports use of parallel requests with curl
```

tklauser commented on cilium-cli/connectivity/check/context.go:
```diff
@@ -945,11 +945,21 @@ func (ct *ConnectivityTest) CurlCommand(peer TestPeer, ipFam features.IPFamily,
-		net.JoinHostPort(peer.Address(ipFam), fmt.Sprint(peer.Port())),
-		peer.Path()))
+
+	for i := 0; i < numTargets; i++ {
```
nit: could use the range-over-int feature [added in Go 1.22](https://go.dev/doc/go1.22#language) here:

```suggestion
	for i := range numTargets {
```

jrajahalme replied:
TIL, thanks!

jrajahalme replied:
Applied!

tklauser reviewed (commented):
Given this is mainly targetted towards stress-testing agent<->proxy interaction, would it make sense to make this conditional on tests using l7 policies?

It also looks like there are some legitimate failures, mainly in tests involving l7 policies.

jrajahalme committed:
envoy: Increase per-cluster concurrent retry limit

Envoy default limit on concurrent retries on a cluster is 3. This has
been seen hit on CI when tests use multiple parallel requests from
curl. Increase the limit to 128 via a new helm value
'envoy.maxConcurrentRetries' and Cilium Agent command line argument
'--proxy-max-concurrent-retries'.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
cilium-cli: Add support for testing with multiple parallel curl requests

Add new 'connectivity test' option '--curl-parallel' which, when set to a
value greater than one causes each curl invocation to issue parallel
requests on concurrent connections. This can be used to add stress to
agent/proxy (DNS and/or Envoy) synchronization.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
helm: Expose '--http-retry-count'

Envoy HTTP retry count defaults to 3. Expose it via helm value
'envoy.httpRetryCount' to make tuning of it easier.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

tklauser commented on cilium-cli/cli/connectivity.go:
```diff
@@ -162,6 +162,7 @@ func newCmdConnectivityTest(hooks api.Hooks) *cobra.Command {
 	cmd.Flags().DurationVar(&params.ConnectTimeout, "connect-timeout", defaults.ConnectTimeout, "Maximum time to allow initiation of the connection to take")
 	cmd.Flags().DurationVar(&params.RequestTimeout, "request-timeout", defaults.RequestTimeout, "Maximum time to allow a request to take")
 	cmd.Flags().BoolVar(&params.CurlInsecure, "curl-insecure", false, "Pass --insecure to curl")
+	cmd.Flags().UintVar(&params.CurlParallel, "curl-parallel", defaults.CurlParallel, "Number of parallel requests in curl commands")
```
Maybe mention in the usage that 0 disables parallel requests?

jrajahalme replied:
Changed usage to:
```
Number of parallel requests in curl commands (0 to disable)
```
